magento/pwa-studio
# Correct dialog form values after submission

## Description

This ticket originally called for the submit / confirm handler to clear out form values like the reset / cancel button does.

Unfortunately, that didn't end up being viable. Take the following code addition to the `Dialog` component itself:

```js
    const onSubmit = useCallback(
        async args => {
            if (onConfirm) await onConfirm(args);
            formApiRef.current.reset();
        },
        [onConfirm]
    );
```

We actually only want to reset the form when `onConfirm` is successful. When this function results in an error, we want to keep the form fields in their current state so that the user can more readily fix the problem.

Unfortunately, there's no established pattern or enforcement of a return value from the `onConfirm` callback. In fact, most of the ones in the Venia storefront return `undefined`, explicitly or implicitly.

It feels very heavy handed and error prone for the `Dialog` component to suggest or require such a return value from its callers.

As such, I think the best approach for now is a targeted one where callers set `shouldUnmountOnHide` to `true` on their `Dialog`s when they want the reset functionality. The form will be re-mounted when the `Dialog` reappears on screen, and the (new, updated)`initialValues` will be used.

To facilitate this, this PR changes the default value for the `Dialog` `shouldUnmountOnHide` prop to `true`.
Any instance of `Dialog` that was explicitly setting `shouldUnmountOnHide` to `true` has been updated to omit that prop, and any instance of `Dialog` that was missing `shouldUnmountOnHide` / relying on the default value has been updated to explicitly set `shouldUnmountOnHide` to `false`.

## Related Issue





Closes PWA-961.

## Acceptance



### Verification Stakeholders

### Specification



### Verification Steps


1. Sign In
2. Navigate to the `/account-information` page (`My Account` > `Account Information`)
3. Click the "Edit" button
4. Change the "Last Name" field and enter the correct password
5. Keep a close eye on the "Last Name" field and click "Submit"
6. Verify that there was no flash of old content in the "Last Name" field
7. Re-open the "Edit" Dialog by clicking on the "Edit" button
8. Verify that the "Last Name" field is correctly updated to the new value and that the password field has been cleared out

## Screenshots / Screen Captures (if appropriate)

![Screen Shot 2021-01-21 at 12 39 51 PM](https://user-images.githubusercontent.com/13182778/105396677-c96e0080-5be5-11eb-8cdc-1ebbba227b76.png)

## Checklist

* I have added tests to cover my changes, if necessary.
* I have added translations for new strings, if necessary.
* I have updated the documentation accordingly, if necessary.

## Risks

Do we need to make this change anywhere else?
